### DISCUSSION TITLE
[Start]: Throwing errors with a status code

### DISCUSSION THREAD

**Discussion by notKamui** [TanStack/router#2887] [Ideas]
Would it be possible to throw errors in server functions / middlewares along with a custom status code instead of systematically respond with a 500 code. For clarity and documentation, always returning a 500 seems a bit strange.

If not that, why can't we abort a request within a middleware by returning a custom response instead of the response from `await next()` ?

In particular, as an example, it would be nice to have a middleware that catches ZodErrors to respond with a 400 and a body containing the validation errors.

**Comment by notKamui** [comment]
I guess it's possible to do this

```ts
throw json({
  message: 'Validation error',
  errors: error.errors,
}, { status: 400 })
```

But `throw`ing the JsonResponse from `json()` (or any Response) does not seem documented, nor really clear/idiomatic

**Reply by vincedelmas** [reply]
I agree, the current approach lacks documentation and practicality. Thank you for providing the alternative solution tho!

**Comment by ImLunaHey** [comment]
is it possible to throw our own custom error classes? would be nice to throw an actual error instead of the `json` method provided.

**Reply by notKamui** [reply]
just wrap the call in a well named function

```ts
function badRequest(data: any, message: string): never {
	throw json({
		data,
		message
	}, { status: 400 })
}
```

**Reply by ImLunaHey** [reply]
that's not at all what i asked. that's a function not an error class.

**Reply by schiller-manuel** [reply]
yes you can do this, you just need a serializationAdapter for it

see https://github.com/TanStack/router/blob/main/e2e/react-start/serialization-adapters/src/CustomError.ts#L18

**Reply by FGoessler** [reply]
Could we please get some proper documentation please for these serialisation adapters? 🙏 

The default serialisation of errors in Tanstack Start server functions is really preventing any details from being passed along. Even the built in validation is barely usable as is as Zod validation errors are not properly serialised.

I was looking for a solution for this for a while and was already about to throw server functions out the window as they seem to immature in regards of testability and error handling. 😬
